File: FCLCore/src/main/java/com/tungsten/fclcore/mod/modinfo/ForgeNewModMetadata.java
package com.tungsten.fclcore.mod.modinfo;

import static com.tungsten.fclcore.util.Logging.LOG;

import com.google.gson.JsonParseException;
import com.moandjiezana.toml.Toml;
import com.tungsten.fclcore.mod.LocalModFile;
import com.tungsten.fclcore.mod.ModLoaderType;
import com.tungsten.fclcore.mod.ModManager;
import com.tungsten.fclcore.util.gson.JsonUtils;
import com.tungsten.fclcore.util.io.CompressingUtils;
import com.tungsten.fclcore.util.io.FileUtils;

import java.io.IOException;
import java.io.InputStream;
import java.nio.file.FileSystem;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardCopyOption;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.jar.Attributes;
import java.util.jar.Manifest;

public final class ForgeNewModMetadata {
    private final String modLoader;

    private final String loaderVersion;

    private final String logoFile;

    private final String license;

    private final List<Mod> mods;

    public ForgeNewModMetadata(String modLoader, String loaderVersion, String logoFile, String license, List<Mod> mods) {
        this.modLoader = modLoader;
        this.loaderVersion = loaderVersion;
        this.logoFile = logoFile;
        this.license = license;
        this.mods = mods;
    }

    public String getModLoader() {
        return modLoader;
    }

    public String getLoaderVersion() {
        return loaderVersion;
    }

    public String getLogoFile() {
        return logoFile;
    }

    public String getLicense() {
        return license;
    }

    public List<Mod> getMods() {
        return mods;
    }

    public static class Mod {
        private final String modId;
        private final String version;
        private final String displayName;
        private final String side;
        private final String displayURL;
        private final String authors;
        private final String description;

        public Mod() {
            this("", "", "", "", "", "", "");
        }

        public Mod(String modId, String version, String displayName, String side, String displayURL, String authors, String description) {
            this.modId = modId;
            this.version = version;
            this.displayName = displayName;
            this.side = side;
            this.displayURL = displayURL;
            this.authors = authors;
            this.description = description;
        }

        public String getModId() {
            return modId;
        }

        public String getVersion() {
            return version;
        }

        public String getDisplayName() {
            return displayName;
        }

        public String getSide() {
            return side;
        }

        public String getDisplayURL() {
            return displayURL;
        }

        public String getAuthors() {
            return authors;
        }

        public String getDescription() {
            return description;
        }
    }

    public static LocalModFile fromForgeFile(ModManager modManager, Path modFile, FileSystem fs) throws IOException {
        return fromFile(modManager, modFile, fs, ModLoaderType.FORGE);
    }

    public static LocalModFile fromNeoForgeFile(ModManager modManager, Path modFile, FileSystem fs) throws IOException {
        return fromFile(modManager, modFile, fs, ModLoaderType.NEO_FORGED);
    }

    private static LocalModFile fromFile(ModManager modManager, Path modFile, FileSystem fs, ModLoaderType modLoaderType) throws IOException {
        if (modLoaderType != ModLoaderType.FORGE && modLoaderType != ModLoaderType.NEO_FORGED) {
            throw new IOException("Invalid mod loader: " + modLoaderType);
        }

        if (modLoaderType == ModLoaderType.NEO_FORGED) {
            try {
                return fromFile0("META-INF/neoforge.mods.toml", modLoaderType, modManager, modFile, fs);
            } catch (Exception ignored) {
            }
        }

        try {
            return fromFile0("META-INF/mods.toml", modLoaderType, modManager, modFile, fs);
        } catch (Exception ignored) {
        }

        try {
            return fromEmbeddedMod(modManager, modFile, fs, modLoaderType);
        } catch (Exception ignored) {
        }

        throw new IOException("File " + modFile + " is not a Forge 1.13+ or NeoForge mod.");
    }

    private static LocalModFile fromFile0(
            String tomlPath,
            ModLoaderType modLoaderType,
            ModManager modManager,
            Path modFile,
            FileSystem fs) throws IOException, JsonParseException {
        Path modToml = fs.getPath(tomlPath);
        if (Files.notExists(modToml))
            throw new IOException("File " + modFile + " is not a Forge 1.13+ or NeoForge mod.");
        Toml toml = new Toml().read(FileUtils.readText(modToml));
        ForgeNewModMetadata metadata = toml.to(ForgeNewModMetadata.class);
        if (metadata == null || metadata.getMods().isEmpty())
            throw new IOException("Mod " + modFile + " `mods.toml` is malformed..");
        Mod mod = metadata.getMods().get(0);
        Path manifestMF = fs.getPath("META-INF/MANIFEST.MF");
        String jarVersion = "";
        if (Files.exists(manifestMF)) {
            try (InputStream is = Files.newInputStream(manifestMF)) {
                Manifest manifest = new Manifest(is);
                jarVersion = manifest.getMainAttributes().getValue(Attributes.Name.IMPLEMENTATION_VERSION);
            } catch (IOException e) {
                LOG.warning("Failed to parse MANIFEST.MF in file " + modFile);
            }
        }

        ModLoaderType type = analyzeLoader(toml, mod.getModId(), modLoaderType);

        return new LocalModFile(modManager, modManager.getLocalMod(mod.getModId(), type), modFile, mod.getDisplayName(), new LocalModFile.Description(mod.getDescription()),
                mod.getAuthors(), jarVersion == null ? mod.getVersion() : mod.getVersion().replace("${file.jarVersion}", jarVersion), "",
                mod.getDisplayURL(),
                metadata.getLogoFile());
    }

    private static LocalModFile fromEmbeddedMod(ModManager modManager, Path modFile, FileSystem fs, ModLoaderType modLoaderType) throws IOException {
        Path manifestFile = fs.getPath("META-INF/MANIFEST.MF");
        if (!Files.isRegularFile(manifestFile))
            throw new IOException("Missing  MANIFEST.MF in file " + manifestFile);

        Manifest manifest;
        try (InputStream input = Files.newInputStream(manifestFile)) {
            manifest = new Manifest(input);
        }

        List<Path> embeddedModFiles = List.of();

        String embeddedDependenciesMod = manifest.getMainAttributes().getValue("Embedded-Dependencies-Mod");
        if (embeddedDependenciesMod != null) {
            Path embeddedModFile = fs.getPath(embeddedDependenciesMod);
            if (!Files.isRegularFile(embeddedModFile)) {
                LOG.warning("Missing embedded-dependencies-mod: " + embeddedModFile);
                throw new IOException();
            }
            embeddedModFiles = List.of(embeddedModFile);
        } else {
            Path jarInJarMetadata = fs.getPath("META-INF/jarjar/metadata.json");
            if (Files.isRegularFile(jarInJarMetadata)) {
                JarInJarMetadata metadata = JsonUtils.fromJsonFile(jarInJarMetadata, JarInJarMetadata.class);
                if (metadata == null)
                    throw new IOException("Invalid metadata file: " + jarInJarMetadata);

                metadata.validate();

                embeddedModFiles = new ArrayList<>();
                if (metadata.getJars() != null) {
                    for (EmbeddedJarMetadata jar : metadata.getJars()) {
                        Path path = fs.getPath(jar.getPath());
                        if (Files.isRegularFile(path)) {
                            embeddedModFiles.add(path);
                        } else {
                            LOG.warning("Missing embedded-dependencies-mod: " + path);
                        }
                    }
                }
            }
        }

        if (embeddedModFiles.isEmpty()) {
            throw new IOException("Missing embedded mods");
        }

        Path tempFile = Files.createTempFile("hmcl-", ".zip");
        try {
            for (Path embeddedModFile : embeddedModFiles) {
                Files.copy(embeddedModFile, tempFile, StandardCopyOption.REPLACE_EXISTING);
                try (FileSystem embeddedFs = CompressingUtils.createReadOnlyZipFileSystem(tempFile)) {
                    return fromFile(modManager, modFile, embeddedFs, modLoaderType);
                } catch (Exception ignored) {
                }
            }
        } finally {
            Files.deleteIfExists(tempFile);
        }

        throw new IOException();
    }

    private static ModLoaderType analyzeLoader(Toml toml, String modID, ModLoaderType loader) throws IOException {
        List<HashMap<String, Object>> dependencies = toml.getList("dependencies." + modID);
        if (dependencies == null) {
            try {
                dependencies = toml.getList("dependencies"); // ??? I have no idea why some of the Forge mods use [[dependencies]]
            } catch (ClassCastException e) {
                try {
                    Toml table = toml.getTable("dependencies");
                    if (table == null)
                        return loader;

                    dependencies = table.getList(modID);
                } catch (Throwable ignored) {
                }
            }

            if (dependencies == null) {
                return loader;
            }
        }

        ModLoaderType result = null;
        loop:
        for (HashMap<String, Object> dependency : dependencies) {
            switch ((String) dependency.get("modId")) {
                case "forge":
                    result = ModLoaderType.FORGE;
                    break loop;
                case "neoforge":
                    result = ModLoaderType.NEO_FORGED;
                    break loop;
            }
        }

        if (result == loader)
            return result;
        else if (result != null)
            throw new IOException("Loader mismatch");
        else {
            LOG.warning("Cannot determine the mod loader for mod " + modID + ", expected " + loader);
            return loader;
        }
    }

}

File: FCL/src/main/java/com/tungsten/fclcore/mod/modinfo/ForgeNewModMetadata.kt
@file:OptIn(kotlinx.serialization.InternalSerializationApi::class)

package com.tungsten.fclcore.mod.modinfo

import com.tungsten.fclcore.mod.LocalModFile
import com.tungsten.fclcore.mod.ModLoaderType
import com.tungsten.fclcore.mod.ModManager
import com.tungsten.fclcore.util.Logging
import com.tungsten.fclcore.util.io.CompressingUtils
import com.tungsten.fclcore.util.io.FileUtils
import kotlinx.serialization.KSerializer
import kotlinx.serialization.Serializable
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonPrimitive
import org.tomlj.Toml
import org.tomlj.TomlParseResult
import org.tomlj.TomlTable
import java.io.IOException
import java.nio.file.FileSystem
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.StandardCopyOption
import java.util.StringJoiner
import java.util.jar.Attributes
import java.util.jar.Manifest

@Serializable
data class ForgeNewModMetadata(
    val modLoader: String = "",
    val loaderVersion: String = "",
    val logoFile: String = "",
    val license: String = "",
    val mods: List<Mod> = emptyList()
) {
    @Serializable
    data class Mod(
        val modId: String = "",
        val version: String = "",
        val displayName: String = "",
        val side: String = "",
        val displayURL: String = "",
        @Serializable(with = ModAuthorsSerializer::class)
        val authors: String = "",
        val description: String = ""
    )

    companion object {
        private val LOG = Logging.LOG

        @JvmStatic
        @Throws(IOException::class)
        fun fromForgeFile(modManager: ModManager, modFile: Path, fs: FileSystem): LocalModFile =
            fromFile(modManager, modFile, fs, ModLoaderType.FORGE)

        @JvmStatic
        @Throws(IOException::class)
        fun fromNeoForgeFile(modManager: ModManager, modFile: Path, fs: FileSystem): LocalModFile =
            fromFile(modManager, modFile, fs, ModLoaderType.NEO_FORGED)

        private fun fromFile(
            modManager: ModManager,
            modFile: Path,
            fs: FileSystem,
            modLoaderType: ModLoaderType
        ): LocalModFile {
            if (modLoaderType != ModLoaderType.FORGE && modLoaderType != ModLoaderType.NEO_FORGED) {
                throw IOException("Invalid mod loader: $modLoaderType")
            }

            if (modLoaderType == ModLoaderType.NEO_FORGED) {
                try {
                    return fromFile0("META-INF/neoforge.mods.toml", modLoaderType, modManager, modFile, fs)
                } catch (ignored: Exception) {
                }
            }

            try {
                return fromFile0("META-INF/mods.toml", modLoaderType, modManager, modFile, fs)
            } catch (ignored: Exception) {
            }

            try {
                return fromEmbeddedMod(modManager, modFile, fs, modLoaderType)
            } catch (ignored: Exception) {
            }

            throw IOException("File $modFile is not a Forge 1.13+ or NeoForge mod.")
        }

        private fun fromFile0(
            tomlPath: String,
            modLoaderType: ModLoaderType,
            modManager: ModManager,
            modFile: Path,
            fs: FileSystem
        ): LocalModFile {
            val modToml = fs.getPath(tomlPath)
            if (Files.notExists(modToml))
                throw IOException("File $modFile is not a Forge 1.13+ or NeoForge mod.")
            val tomlParseResult: TomlParseResult = Toml.parse(FileUtils.readText(modToml))
            if (tomlParseResult.hasErrors()) {
                val ioException = IOException("Mod $modFile $tomlPath is malformed..")
                tomlParseResult.errors().forEach(ioException::addSuppressed)
                throw ioException
            }
            val metadata = MOD_METADATA_JSON.decodeFromString<ForgeNewModMetadata>(tomlParseResult.toJson())
            if (metadata.mods.isEmpty())
                throw IOException("Mod $modFile $tomlPath is malformed..")
            val mod = metadata.mods.first()
            val manifestMF = fs.getPath("META-INF/MANIFEST.MF")
            var jarVersion: String? = ""
            if (Files.exists(manifestMF)) {
                try {
                    Files.newInputStream(manifestMF).use { input ->
                        jarVersion = Manifest(input).mainAttributes.getValue(Attributes.Name.IMPLEMENTATION_VERSION)
                    }
                } catch (e: IOException) {
                    LOG.warning("Failed to parse MANIFEST.MF in file $modFile")
                }
            }

            val type = analyzeLoader(tomlParseResult, mod.modId, modLoaderType)

            return LocalModFile(
                modManager, modManager.getLocalMod(mod.modId, type), modFile, mod.displayName,
                LocalModFile.Description(mod.description), mod.authors,
                if (jarVersion != null) mod.version.replace("\${file.jarVersion}", jarVersion) else mod.version,
                "",
                mod.displayURL,
                metadata.logoFile
            )
        }

        private fun fromEmbeddedMod(
            modManager: ModManager,
            modFile: Path,
            fs: FileSystem,
            modLoaderType: ModLoaderType
        ): LocalModFile {
            val manifestFile = fs.getPath("META-INF/MANIFEST.MF")
            if (Files.notExists(manifestFile))
                throw IOException("Missing MANIFEST.MF in file $modFile")

            val manifest: Manifest = Files.newInputStream(manifestFile).use { Manifest(it) }

            var embeddedModFiles: List<Path> = emptyList()

            val embeddedDependenciesMod = manifest.mainAttributes.getValue("Embedded-Dependencies-Mod")
            if (embeddedDependenciesMod != null) {
                val embeddedModFile = fs.getPath(embeddedDependenciesMod)
                if (Files.notExists(embeddedModFile)) {
                    LOG.warning("Missing embedded-dependencies-mod: $embeddedDependenciesMod")
                    throw IOException()
                }
                embeddedModFiles = listOf(embeddedModFile)
            } else {
                val jarInJarMetadata = fs.getPath("META-INF/jarjar/metadata.json")
                if (Files.exists(jarInJarMetadata)) {
                    val metadata: JarInJarMetadata =
                        MOD_METADATA_JSON.decodeFromString(FileUtils.readText(jarInJarMetadata))
                    embeddedModFiles = ArrayList()
                    for (jar in metadata.jars) {
                        val path = fs.getPath(jar.path)
                        if (Files.exists(path)) {
                            embeddedModFiles += path
                        } else {
                            LOG.warning("Missing embedded-dependencies-mod: ${jar.path}")
                        }
                    }
                }
            }

            if (embeddedModFiles.isEmpty()) {
                throw IOException("Missing embedded mods")
            }

            val tempFile = Files.createTempFile("hmcl-", ".zip")
            try {
                for (embeddedModFile in embeddedModFiles) {
                    Files.copy(embeddedModFile, tempFile, StandardCopyOption.REPLACE_EXISTING)
                    try {
                        CompressingUtils.createReadOnlyZipFileSystem(tempFile).use { embeddedFs ->
                            return fromFile(modManager, modFile, embeddedFs, modLoaderType)
                        }
                    } catch (ignored: Exception) {
                    }
                }
            } finally {
                Files.deleteIfExists(tempFile)
            }

            throw IOException()
        }

        private fun analyzeLoader(toml: TomlParseResult, modID: String, loader: ModLoaderType): ModLoaderType {
            var dependencies: List<Map<String, Any>>? = null
            try {
                val tomlArray = toml.getArray("dependencies.$modID")
                if (tomlArray != null) {
                    dependencies = tomlArray.toList().map { (it as TomlTable).toMap() }
                }
            } catch (ignored: ClassCastException) { // https://github.com/HMCL-dev/HMCL/issues/5068
            }

            if (dependencies == null) {
                try {
                    val tomlArray = toml.getArray("dependencies") // ??? I have no idea why some of the Forge mods use [[dependencies]]
                    if (tomlArray != null) {
                        dependencies = tomlArray.toList().map { (it as TomlTable).toMap() }
                    }
                } catch (e: ClassCastException) {
                    try {
                        val table = toml.getTable("dependencies") ?: return loader
                        val tomlArray = table.getArray(modID)
                        if (tomlArray != null) {
                            dependencies = tomlArray.toList().map { (it as TomlTable).toMap() }
                        }
                    } catch (ignored: Throwable) {
                    }
                }

                if (dependencies == null) {
                    return loader
                }
            }

            var result: ModLoaderType? = null
            loop@ for (dependency in dependencies) {
                when (dependency["modId"] as? String) {
                    "forge" -> {
                        result = ModLoaderType.FORGE
                        break@loop
                    }
                    "neoforge" -> {
                        result = ModLoaderType.NEO_FORGED
                        break@loop
                    }
                }
            }

            return if (result != null) {
                if (result != loader)
                    LOG.warning("Loader mismatch for mod $modID, found $result, expecting $loader")
                result
            } else {
                LOG.warning("Cannot determine the mod loader for mod $modID, expected $loader")
                loader
            }
        }
    }
}

/** mods.toml 的 authors 兼容字符串与字符串数组两种 JSON 形态 */
internal object ModAuthorsSerializer : KSerializer<String> {
    private val elementSerializer = JsonElement.serializer()

    override val descriptor: SerialDescriptor = elementSerializer.descriptor

    override fun deserialize(decoder: Decoder): String {
        val element = decoder.decodeSerializableValue(elementSerializer)
        return when (element) {
            is JsonPrimitive -> element.content
            is JsonArray -> {
                val joiner = StringJoiner(", ")
                for (child in element) {
                    if (child !is JsonPrimitive) {
                        return element.toString()
                    }
                    joiner.add(child.content)
                }
                joiner.toString()
            }
            else -> element.toString()
        }
    }

    override fun serialize(encoder: Encoder, value: String) {
        encoder.encodeSerializableValue(elementSerializer, JsonPrimitive(value))
    }
}

/** Jar-in-Jar 嵌入依赖元数据（META-INF/jarjar/metadata.json） */
@Serializable
internal data class JarInJarMetadata(val jars: List<EmbeddedJarMetadata> = emptyList())

@Serializable
internal data class EmbeddedJarMetadata(val path: String = "", val isObfuscated: Boolean = false)
